### DISCUSSION TITLE
Is it possible to use the skia renderer in software mode for embedded?

### DISCUSSION THREAD

**Discussion by FoundationKen** [slint-ui/slint#3782] [General]
We've been building a UI for an embedded project, but have been doing the design on desktop up until now.  We just started testing the slint files on our embedded platform and ran into some issues, which we didn't realize would be problem before.

1. There are no drop shadows for rectangles.
2. Rectangles with rounded corners do not clip to the rounded corners (only to the squared-off corners).
3. Fonts are not rendered dynamically, but you must export the required sizes.
4. Path/LineTo/ArcTo components don't work, but SVGs do.
5. Asian characters don't render (maybe just my font?).
6. Rendering speed is VERY slow, so animations are very choppy.

We switched our desktop app as shown below to make it match the platform's current renderer:

```
[dependencies.slint]
version = "1.2.2"
features = ["backend-winit", "renderer-winit-software", "compat-1-2"]
default-features = false
```

Now, we want to try the skia renderer to see if it fixes some of these issues, but with a software surface.  Is there a way to configure this at the moment?

Thanks!

**Comment by tronical** [comment]
Yes, we recently added support for rendering with Skia's software renderer. It's not optimised yet in the sense that there are no partial updates yet, but that can be implemented. If you want to give it a try, check out Slint from git's master branch, add `"renderer-skia"` to your feature flags and set `SLINT_BACKEND=winit-skia-software` as environment variable. Alternatively, if you run this on a system without OpenGL/Metal/Direct3D, then the skia renderer now automatically falls back to this software rendering code path.

What kind of device / screen resolution are you targeting? Even with the built-in software renderer animations should not be sloppy, so I'm very curious :)

**Reply by FoundationKen** [reply]
Following up via Slint chat.

**Comment by eupn** [comment]
Hi @tronical, is there a way to use software skia rendering on an embedded platform? Currently I'm running into shared fonts issues as there's no shared libraries on the embedded platform.

```rust
error[E0433]: failed to resolve: could not find `Library` in `libloading`
  --> internal/common/sharedfontdb/fontconfig.rs:21:38
   |
21 |         let fontconfig = libloading::Library::new("libfontconfig.so.1")?;
   |                                      ^^^^^^^ could not find `Library` in `libloading`
   ```

Maybe we'd need to use the same glyph embedding technique and render the fonts with `fontdue` as it's done for the software renderer? But at the same time use skia to draw everything else (again in software)?

**Reply by tronical** [reply]
Could you describe the embedded platform that you’re using that doesn’t support shared libraries?

In theory what you’re asking is possible, but I’m very curious how you’d go about compiling Skia on the platform. Out of the box on Linux it requires fontfonfig as shared library.

**Reply by eupn** [reply]
@tronical it's our custom embedded platform & OS and it has rust `std` support. It doesn't support shared libraries or any standard font sharing mechanisms, which means we'd have to use `fontdue` or similar for now. We're currently using slint's software renderer (`MinimalSoftwareWindow`) with no issues (albeit with less supported features) but wanted to use `skia` in the same manner.

**Reply by tronical** [reply]
Thanks for the info. In that case I think that if you port Skia and [rust-skia](https://github.com/rust-skia/rust-skia) to basically only include skia’s software renderer, then there might be ways of patching Slint to use that in conjunction with pre-rendered glyphs. (Technically speaking :-)
